python/ray/serve/api.py

```diff
@@ -1510,155 +1516,394 @@ def get_deployment_statuses() -> Dict[str, DeploymentStatusInfo]:
+            "Class used in DAG should not be in-line defined when exporting"
+            "import path for deployment. Please ensure it has fully "
+            "qualified name with valid __module__ and __qualname__ for "
+            "import path, with no __main__ or <locals>. \n"
+            f"Current import path: {import_path}"
+        )
+        assert "__main__" not in import_path, error_message
+        assert "<locals>" not in import_path, error_message
 
-        Returns:
-            Dict: The Application's deployments formatted in a dictionary.
-        """
-        raise NotImplementedError()
+        json_dict["import_path"] = import_path
+        json_dict["deployment_config"] = base64.b64encode(
+            self._deployment.config.to_proto_bytes()
+        ).decode()
 
-    @classmethod
-    def from_dict(cls, d: Dict) -> "Application":
-        """Converts a dictionary of deployment data to an application.
+        return json_dict
 
-        Takes in a dictionary matching the Serve REST API schema and converts
-        it to an application containing those deployments.
+    @classmethod
+    def from_json(cls, input_json, object_hook=None):
+        assert input_json[DAGNODE_TYPE_KEY] == DeploymentNode.__name__
+        args_dict = super().from_json_base(input_json, object_hook=object_hook)
+        deployment_config = DeploymentConfig.from_proto_bytes(
+            base64.b64decode(input_json["deployment_config"])
+        )
+        return cls(
+            input_json["import_path"],
+            input_json["deployment_name"],
+            args_dict["args"],
+            args_dict["kwargs"],
+            args_dict["options"],
+            deployment_config,
+            other_args_to_resolve=args_dict["other_args_to_resolve"],
+        )
 
-        Args:
-            d (Dict): A dictionary containing the deployments' data that matches
-                the Serve REST API schema.
 
-        Returns:
-            Application: a new application object containing the deployments.
-        """
-        raise NotImplementedError()
+class UnboundDeploymentMethodNode(DAGNode):
+    def __init__(
+        self,
+        deployment: Deployment,
+        method_name: str,
+        other_args_to_resolve: Dict[str, Any] = None,
+    ):
+        self._deployment = deployment
+        self._method_name = method_name
+        self._options = {}
+        self._bound_other_args_to_resolve = other_args_to_resolve
+
+    def bind(self, *args, **kwargs):
+        node = DeploymentMethodNode(
+            self._deployment,
+            self._method_name,
+            args,
+            kwargs,
+            self._options,
+            other_args_to_resolve=self._bound_other_args_to_resolve,
+        )
+        return node
 
-    def to_yaml(self, f: Optional[TextIO] = None) -> Optional[str]:
-        """Returns this application's deployments as a YAML string.
+    def options(self, **options):
+        self._options = options
```

Comment:
should be `self._options.update(options)`, right? otherwise we can't set partial options.